papermill/execute.py

```diff
@@ -13,26 +15,75 @@
 from six import string_types
 
 
-class PapermillExecutePreprocessor(ExecutePreprocessor):
+def preprocess(self, nb, resources):
+    """
+    This function is for monkey patching the nbconvert Preprocessor.preprocess method.
+    We are doing this for the following reasons:
+    1. Notebooks will stop executing when they encounter a failure but not raise a CellException.
+       This allows us to save the notebook with the traceback even though a CellExecutionError
+       was encountered.
+
+    2. We want to write the notebook as cells are executed. We inject our logic for that here.
 
-    def preprocess(self, nb, resources):
-        """
-        The default preprocess behavior if allow_errors = True is to continue executing all the cells.
-        We want the notebook to cease execution and then write out the state of the notebook with the traceback
-        on the cell that failed.
-        """
-        # Clear out all outputs prior to execution.
-        for cell in nb.cells:
-            cell.outputs = []
+    3. We will be included cell execution metrics.
 
-        # Catch CellExecutionError if thrown which lets this method finish so we can write the state of the notebook
-        # to disk.
-        try:
-            nb, resources = super(PapermillExecutePreprocessor, self).preprocess(nb, resources)
-        except CellExecutionError:
-            pass
+    Original Docstring:
 
-        return nb, resources
+    Preprocessing to apply on each notebook.
+
+    Must return modified nb, resources.
+
+    If you wish to apply your preprocessing to each cell, you might want
+    to override preprocess_cell method instead.
+
+    Parameters
+    ----------
+    nb : NotebookNode
+        Notebook being converted
+    resources : dictionary
+        Additional resources used in the conversion process.  Allows
+        preprocessors to pass variables into the Jinja engine.
+    """
+    output_path = nb.metadata.papermill['output_path']
+
+    # Reset the notebook.
+    for cell in nb.cells:
+        # Reset the cell execution counts.
+        if hasattr(cell, "execution_count"):
+            cell.execution_count = None
+
+        # Clear out the papermill metadata for each cell.
+        cell.metadata['papermill'] = dict(
+            exception=None,
+            start_time=None,
+            end_time=None,
+            duration=None
+        )
+        cell.outputs = []
+
+    # Execute each cell and update the output in real time.
+    writers = []
+    with futures.ThreadPoolExecutor(max_workers=1) as executor:
+        for index, cell in enumerate(nb.cells):
+            cell.execution_count = "*"
+            writers.append(executor.submit(write_ipynb, nb, output_path))
```

Comment:
I want to write asynchronously to not add more time to running a cell than I have to.